qubes/ext/block.py

```diff
@@ -105,22 +113,168 @@ def size(self):
 
     @property
     def device_node(self):
-        '''Device node in backend domain'''
+        """Device node in backend domain"""
         return '/dev/' + self.ident.replace('_', '/')
 
+    @property
+    def interfaces(self) -> List[qubes.devices.DeviceInterface]:
+        """
+        List of device interfaces.
+
+        Every device should have at least one interface.
+        """
+        return [qubes.devices.DeviceInterface("******", "block")]
+
+    @property
+    def parent_device(self) -> Optional[qubes.devices.Device]:
+        """
+        The parent device if any.
+
+        If the device is part of another device (e.g. it's a single
+        partition of an usb stick), the parent device id should be here.
+        """
+        if self._parent is None:
+            if not self.backend_domain.is_running():
+                return None
+            untrusted_parent_info = self.backend_domain.untrusted_qdb.read(
+                f'/qubes-block-devices/{self.ident}/parent')
+            if untrusted_parent_info is None:
+                return None
+            else:
+                # '4-4.1:1.0' -> parent_ident='4-4.1', interface_num='1.0'
+                # 'sda' -> parent_ident='sda', interface_num=''
+                parent_ident, sep, interface_num = self._sanitize(
+                    untrusted_parent_info).partition(":")
+                devclass = 'usb' if sep == ':' else 'block'
+                self._parent = qubes.devices.Device(
+                    self.backend_domain, parent_ident, devclass=devclass)
+                self._interface_num = interface_num
+        return self._parent
+
+    @property
+    def self_identity(self) -> str:
+        """
+        Get identification of device not related to port.
+        """
+        parent_identity = ''
+        p = self.parent_device
+        if p is not None:
+            p_info = p.backend_domain.devices[p.devclass][p.ident]
+            parent_identity = p_info.self_identity
+            if p.devclass == 'usb':
+                parent_identity = f'{p.ident}:{parent_identity}'
+        if self._interface_num:
+            # device interface number
+            self_id = self._interface_num
+        else:
+            # partition number: 'sda1' -> '1'
+            self_id = self.ident[3:]
```

Comment:
That wont always work, there can be other names (`xvdi1`), or even different schemes (`nvme0n1p2`). Generally, to get partition number, take last number (not just last digit, there can be `sda12`). But there is still one corner case - whole device ending with a number, like `nvme0n1`, or `loop0` - in those cases the partition number is separated with the `p`, but you can't tell it's a partition just because it ends with a number... Does the partition device always has "whole block device" as a parent? If so, you can use that fact and only then take partition number. But if parent device is not a block device, then it isn't a partition, so getting just the final number is not appropriate.

Reply:
Youre right, I fixed it